geomesa-spring/geomesa-spring-cloud-stream-binder-kafka-datastore/src/main/java/org/locationtech/geomesa/spring/binder/kafka/datastore/KafkaDatastoreMessageProducer.java

```diff
@@ -0,0 +1,70 @@
+/***********************************************************************
+ * Copyright (c) 2013-2024 Commonwealth Computer Research, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Apache License, Version 2.0
+ * which accompanies this distribution and is available at
+ * http://www.opensource.org/licenses/apache2.0.php.
+ ***********************************************************************/
+
+package org.locationtech.geomesa.spring.binder.kafka.datastore;
+
+
+import org.geotools.api.data.DataStore;
+import org.geotools.api.data.FeatureEvent;
+import org.geotools.api.data.SimpleFeatureStore;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.cloud.stream.provisioning.ConsumerDestination;
+import org.springframework.integration.endpoint.MessageProducerSupport;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.support.MessageBuilder;
+
+import java.io.IOException;
+import java.util.function.Supplier;
+
+public class KafkaDatastoreMessageProducer extends MessageProducerSupport {
+    private static final Logger logger = LoggerFactory.getLogger(KafkaDatastoreMessageProducer.class);
+
+    private final ConsumerDestination destination;
+    private final Supplier<DataStore> dsFactory;
+    private DataStore ds;
+
+    public KafkaDatastoreMessageProducer(ConsumerDestination destination,
+                                         Supplier<DataStore> dsFactory) {
+        this.destination = destination;
+        this.dsFactory = dsFactory;
+    }
+
+    @Override
+    public void doStart() {
+        SimpleFeatureStore fs = getSimpleFeatureStore();
+
+        fs.addFeatureListener(featureEvent -> {
```

Comment:
so feature listeners don't guarantee at-least-once delivery - if you want that, you need to use a GeoMessageProcessor with a stable group id: https://github.com/locationtech/geomesa/blob/main/geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/data/KafkaDataStore.scala#L124

I'm realizing that method is not in the docs... I'll open a ticket to add it there.